ddstdecode.1
'\" t
'\"
'\"	DO NOT EDIT! This file is generated from ddstdecode.1in
'\"
.TH ddstdecode 1 "Sun Apr 25 16:22:30 2021" "ddstdecode 0.0"
'\"
'\"==========================================================================
'\"	STRINGS and MACROS
'\"==========================================================================
'\"
'\"	Define strings for special characters that nroff doesn't have
'\"
'\"	N.B. using .if cua to test for special characters did not work.
'\"
.if !'\*[.T]'ps' .ds ua up
.if  '\*[.T]'ps' .ds ua \(ua
.if !'\*[.T]'ps' .ds da down
.if  '\*[.T]'ps' .ds da \(da
.if !'\*[.T]'ps' .ds <- left
.if  '\*[.T]'ps' .ds <- \(<-
.if !'\*[.T]'ps' .ds -> right
.if  '\*[.T]'ps' .ds -> \(->
'\"
'\"     bx - box a Courier string for making keycaps
'\"
'\"	N.B. this mess is to make the line drawing come out only
'\"	if we are really generating postscript
'\"
.de bx
.ie !'\*[.T]'ps' \{\
.	RB [ \\$1 ]\\$2
.\}
.el \{\
.	ie !r ps4html \{\
'\" \(br\|\s-1\f(CB\\$1\fP\s+1\|\(br\l'|0\(rn'\l'|0\(ul'
.		ft CW
.		nr par*bxw \w'\\$1'+.4m 
\Z'\v'.25m'\D'l 0 -1m'\D'l \\n[par*bxw]u 0'\D'l 0 1m'\D'l -\\n[par*bxw]u 0''\
\Z'\h'.2m'\s-1\\$1\s+1'\
\h'\\n[par*bxw]u'\\$2
.		ft P
.	\}
.	el \{\
.		RB [ \\$1 ]\\$2
.	\}
.\}
..
'\"
'\" strings to set current color (null with old groff)
'\"
.if mred .ds red \m[red]
.if mgreen .ds green \m[green]
.if mblue .ds blue \m[blue]
.if mblack .ds black \m[black]
.if mblack .ds mP \mP
'\"
'\" fix for grotty + xterm. We call for orange, grotty outputs yellow,
'\" but xterm displays yellow as orange.  The cycle is complete.
'\"
.if n .defcolor orange rgb #ffff00
'\"
'\" color <color> - set the current color (ignores request with old groff)
'\"
.de color
.if mred \m[\\$1]\c
..
'\"
'\" colorword <color> <word> - colorize a word (ignored by old groff)
'\"
.de colorword
.ie m\\$1 \m[\\$1]\\$2\mP\c
.el \\$2\c
..
'\"
'\" colbox <fg> <bg> <word> - colorize a word in a filled box
'\"
.de colbox
.ie mred \M[\\$2]\
\v'+.167v'\
\D'P 0 -0.9v  \w'\\$3'u 0  0 +0.9v   -\w'\\$3'u 0'\
\v'-.167v'\
\m[\\$1]\\$3\mP\MP
.el \\$3\c
..
'\"
'\"	Macros for doing pdfmarks
'\"
.de specialps
.if  '\*[.T]'ps' \\k_\X'ps: \\$*'\h'|\\n_u'\c
..
'\"
'\" pdfmark PDFMARKCODE
'\"
.ds pdfmarks
.if d pdfmarks \{\
.de pdfmark
.	specialps exec [\\$1 pdfmark
..
'\"
'\" pdfdest LINKNAME
'\"
.de pdfdest
.pdfmark "/Dest /\\$1 /View [/XYZ -5 PL null] /DEST"
..
'\"
'\" pdfbookmark COUNT LINKNAME STRING
'\"
.de pdfbookmark
.   pdfmark "/View [/XYZ 44 730 1.0] /Count \\$1 /Dest /\\$2 /Title (\\$3) /OUT"
..
'\"
'\"	Define the SH and SS macros to save pdfmark information
'\"	in "arrays" of numbers and strings.
'\"
.if !r rr_n \{\
.nr rr_n 0 1
.am SH
.	nr rr_levels!\\n+[rr_n] 2
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.am SS
.	nr rr_levels!\\n+[rr_n] 3
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.\}
'\"
'\"	Called at the end of the document to generate the pdfmark outline
'\"
.de pdf_outline
.nr rr_levels!\\n+[rr_n] 1
.nr rr_i 0 1
.while \\n+[rr_i]<\\n[rr_n] \{\
.   nr rr_ip1 \\n[rr_i]+1
.   nr rr_count 0
.   if \\n[rr_levels!\\n[rr_ip1]]>\\n[rr_levels!\\n[rr_i]] \{\
.       nr rr_j \\n[rr_i] 1
.       while \\n+[rr_j]<\\n[rr_n] \{\
.           if \\n[rr_levels!\\n[rr_j]]<=\\n[rr_levels!\\n[rr_i]] \{\
.               break
.           \}
.           if \\n[rr_levels!\\n[rr_j]]==(\\n[rr_levels!\\n[rr_i]]+1) \{\
.               nr rr_count \\n[rr_count]+1
.           \}
.       \}
.   \}
.   ds hhh \\*[rr_labels!\\n[rr_i]]
.   pdfbookmark -\\n[rr_count] Link\\n[rr_i] "\\*[hhh]"
.\}
..
'\"
'\" Some postscript to make pdfmarks harmless on old interpreters...
'\"
.specialps "def /pdfmark where {pop} {userdict /pdfmark /cleartomark load put} ifelse"
'\"
'\" Force display of Bookmarks in Acrobat when document is viewed.
'\"
.pdfmark "[/PageMode /UseOutlines /Page 1 /View [/XYZ null null null] /DOCVIEW"
'\"
'\" Output the document info in pdfmarks
'\"
.pdfmark "\
	/Title (\*[an-title](\*[an-section])) \
	/Subject (\*[an-title] Manual Page) \
	/Author (Rick Richardson) \
	/Keywords (printing) \
	/Creator (groff \n(.x.\n(.y.\n(.Y -man) \
	/CreationDate (\*[an-extra1]) \
	/ModDate (\*[an-extra1]) \
	/DOCINFO"
\}
'\"
'\" The manual page name is only 1st level mark
'\"
.nr rr_levels!\n+[rr_n] 1
.ds rr_labels!\n[rr_n] \*[an-title](\*[an-section])
.pdfdest Link\n[rr_n]
'\"
'\"==========================================================================
'\"	MANUAL PAGE SOURCE
'\"==========================================================================
.SH NAME
ddstdecode \- Decode a Ricoh DDST stream into human readable form.
.SH SYNOPSIS
.B ddstdecode
.RI [ options "] <" ddst-file
.SH DESCRIPTION
.B ddstdecode
decodes a Ricoh DDST stream into human readable form.
.P
A Ricoh DDST stream is the printer language used by some Ricoh
printers. From what I can tell, it is pbmtojbg(1) wrapped with some PJL.

.SH COMMAND LINE OPTIONS
These are the options that can appear on the command line.
.TP
.BI \-d\0 basename
Basename of .pbm file for saving decompressed planes.
.TP
.BI \-h
Print hex file offsets.
.TP
.BI \-o
Print file offsets.
.TP
.BI \-D\0 level
Set Debug level [0].

.SH EXAMPLES
Decode an Ricoh DDST stream file created by a Ricoh SP112.

.nf
.ft CW
$ ddstdecode -h ~/testpage-ricoh-sp112.prn
     0: \033%-12345X@PJL
     f: @PJL SET TIMESTAMP=2017/03/16 16:53:16
    37: @PJL SET FILENAME=testpage.pdf
    57: @PJL SET COMPRESS=JBIG
    6f: @PJL SET USERNAME=rick
    87: @PJL SET COVER=OFF
    9b: @PJL SET HOLD=OFF
    ae: @PJL SET PAGESTATUS=START
    c9: @PJL SET COPIES=1
    dc: @PJL SET MEDIASOURCE=TRAY1
    f8: @PJL SET MEDIATYPE=PLAINRECYCLE
   119: @PJL SET PAPER=LETTER
   130: @PJL SET PAPERWIDTH=5100
   14a: @PJL SET PAPERLENGTH=6600
   165: @PJL SET RESOLUTION=600
   17e: @PJL SET IMAGELEN=60604
   197: DDST_JBIG_DATA_BEGIN 60604 bytes
                DL = 0, D = 0, P = 1, - = 0, XY = 5100 x 6600
                L0 = 128, MX = 0, MY = 0
                Order   = 3  ILEAVE SMID
                Options = 72  LRLTWO TPBON
                52 stripes, 0 layers, 1 planes
  ee53: DDST_JBIG_DATA_END
  ee53: @PJL SET DOTCOUNT=1765571
  ee6e: @PJL SET PAGESTATUS=END
  ee87: @PJL EOJ
  ee91: \033%-12345X
.ft P
.fi

.SH FILES
.BR /usr/bin/ddstdecode
.SH SEE ALSO
.BR foo2ddst-wrapper (1),
.BR foo2ddst (1),
.BR pbmtojbg (1)
.SH "AUTHOR"
Rick Richardson <rick.richardson@comcast.net>
.br
http://foo2xqx.rkkda.com/
'\"
'\"
'\"
.em pdf_outline


ChangeLog
2020-06-10	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: Add PREFIX=/usr or PREFIX=/usr/local to cc rules

2020-05-05:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: SaveToner implemented for color pages,
	  thanks to bedi@3mtech.sk .

2020-04-26:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug with SaveToner

2020-02-07:	Rick Richardson <rick.richardson@comcast.net>
	* Add Cassette 2 to foo2oak-z1-InputSlot.xml (KM-1635)

2019-09-09:	Rick Richardson <rick.richardson@comcast.net>
	* Add SmartInstallDisable-Tool.run since it was gone from
	      http://hplipopensource.com/hplip-web/smartinstall/
	  See HP SMART INSTALL NOTES in the INSTALL doc.

2019-04-13:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: Add install-aa to disable AppArmor on SUSE

2018-09-10:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:	revert the 09-06 change.

2018-09-06:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:
	    Use the older foomatic-ppdfile which only has postscript,
	    not postscript and pdf, for making ppd's.

2017-12-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printers: HP LaserJet Pro M12a, M12w

2017-03-20	Rick Richardson <rick.richardson@comcast.net>
	* Add ddstdecode, for decoding some Ricoh printers.

2016-10-21	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: install-hotplug-prog
	    Add /lib/systemd/systemd-udevd to the list for Ubuntu 16.

2016-09-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3040 using the foo2hbpl2 driver.

2016-08-30	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3010 using the foo2hbpl2 driver.

2016-03-13	Rick Richardson <rick.richardson@comcast.net>
	* icc2ps/lcms.h: USE_CUSTOM_SWAB defined.

2015-10-24	Rick Richardson <rick.richardson@comcast.net>
	* foomatic-db/printer/Epson-AcuLaser_CX17NF.xml:
		Change margins on Epson AcuLaser CX17NF.

2015-10-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Epson AcuLaser CX17NF using the foo2hbpl2 driver.

2015-07-04	pavel.shvagirev	<pavel.shvagirev@gmail.com>
	* Modify: hplj10XX to not call us twice on Ubuntu 14.

2015-05-11	Rick Richardson <rick.richardson@comcast.net>
	* Bug in foomatic-db/printer/Xerox-Phaser_6115MFP.xml . Should use
	  foo2lava driver. Thanks to user "ophers".

2015-02-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Oki C511 using foo2hiperc.

2014-12-21	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: make it work with OS X 10.10

2014-12-04	Rick Richardson <rick.richardson@comcast.net>
	* make cups: add "Sandboxing Relaxed" to cups-files.conf iff cups >= 2

2014-09-25	Rick Richardson <rick.richardson@comcast.net>
	* foo2hblp2, foo2hblp2-wrapper, foomatic et al:
	    Add paper size A5.
	* foo2hblp2 foomatic db entries:
	    Changed resolution to 600x600, 1200x600, and 1200x1200.
	    Note: need to recreate the printers if you want more choices!

2014-05-19	Rick Richardson <rick.richardson@comcast.net>
	* HP CLJ CP1215:  Adjust margins: *HWMargins: 18 14.40 18 14

2014-05-04	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2-wrapper: Change PAPER=21 to PAPER=0 for custom page sizes.

2014-04-17	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Fuji Xerox DocuPrint CM215 MFP using the foo2hbpl2
	driver.

2014-04-08	Rick Richardson <rick.richardson@comcast.net>
	* Upgrade to JBIGKIT 2.1

2014-04-01	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2 foomatic-db files and PPDs:
	    Bug with default PageSize.
	    PLEASE delete and recreate the foo2hbpl2 printers.

2014-03-22	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: another redundancy from Dave Coffin

2014-03-18	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More envelope sizes.

2014-03-17	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug, was using A4(1) instead of letter(4) by default
	* hbpldecode: Print out paper codes.

2014-03-15	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More changes from Dave.  Still gets 
	    Segmentation fault (core dumped)
	* hbpldecode: Changed unsigned to int in 2 places.  Works good!!!

2014-03-13	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: Bug with no -d/-r.

2014-03-13	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: Changes to pick apart version 1 from
	    Dave Coffin <dcoffin@cybercom.net>
	* I get:
	    Decoding page 1 to page-01.ppm ...
	    Segmentation fault (core dumped)

2014-03-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Fuji Xerox DocuPrint M215 MFP using the foo2hbpl2 driver.

2014-02-09	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox WorkCentre 3045 MFP using the foo2hbpl2 driver.

2014-01-26	Rick Richardson <rick.richardson@comcast.net>
	* foo2qpdl et al: with -z3 (e.g. CLP-365) you can adjust
	    Brightness and Contrast for monochrome prints, and you can adjust
	    Saturation, Cyan-Red, Magenta-Green, Yellow-Blue Balances.
	* foo2hp: somewhere a bug in foomatic (probably 4.x) crept in
	    so foo2hp-AlignCMYK.xml was only processing one option instead
	    of three options!  Split the file into foo2hp-AlignC.xml,
	    foo2hp-AlignM.xml, and foo2hp-AlignY.xml.

2014-01-24	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Dell C1765nf/nfw using the foo2hbpl2 driver.

2014-01-08	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Epson AcuLaser M1400 using the foo2hbpl2 driver.

2013-12-25	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: HP LaserJet Pro M1212nf MFP using the foo2xqx driver.

2013-12-01	Rick Richardson <rick.richardson@comcast.net>
	* foo2hiperc: clean up man pages

2013-11-29	Rick Richardson <rick.richardson@comcast.net>
	* foo2hiperc-wrapper, foomatic and PPD stuff:
	    add -z1 for Oki C810 so everything will be

ddstdecode.c
/*
 * $Id: ddstdecode.c,v 1.7 2017/03/25 15:01:32 rick Exp $
 */

/*b
 * Copyright (C) 2003-2006  Rick Richardson
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * Author: Rick Richardson <rick.richardson@comcast.net>
b*/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#include "ddst.h"
#include "jbig.h"

/*
 * Global option flags
 */
int	Debug = 0;
char	*DecFile;
int	PrintOffset = 0;
int	PrintHexOffset = 0;

void
debug(int level, char *fmt, ...)
{
    va_list ap;

    if (Debug < level)
	return;

    va_start(ap, fmt);
    vfprintf(stderr, fmt, ap);
    va_end(ap);
}

int
error(int fatal, char *fmt, ...)
{
	va_list ap;

	fprintf(stderr, fatal ? "Error: " : "Warning: ");
	if (errno)
	    fprintf(stderr, "%s: ", strerror(errno));
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	if (fatal > 0)
	    exit(fatal);
	else
	{
	    errno = 0;
	    return (fatal);
	}
}

void
usage(void)
{
    fprintf(stderr,
"Usage:\n"
"	ddstdecode [options] < zjs-file\n"
"\n"
"	Decode a Ricoh DDST stream into human readable form.\n"
"\n"
"	A Ricoh DDST stream is the printer language used by some Ricoh\n"
"	printers. From what I can tell, it is pbmtojbg wrapped with some PJL.\n"
"\n"
"\n"
"Options:\n"
"       -d basename Basename of .pbm file for saving decompressed planes\n"
"       -o          Print file offsets\n"
"       -h          Print hex file offsets\n"
"       -D lvl      Set Debug level [%d]\n"
    , Debug
    );

    exit(1);
}

void
print_bih(unsigned char bih[20])
{
    unsigned int xd, yd, l0;

    xd = (bih[4] << 24) | (bih[5] << 16) | (bih[6] << 8) | (bih[7] << 0);
    yd = (bih[8] << 24) | (bih[9] << 16) | (bih[10] << 8) | (bih[11] << 0);
    l0 = (bih[12] << 24) | (bih[13] << 16) | (bih[14] << 8) | (bih[15] << 0);

    printf("		DL = %d, D = %d, P = %d, - = %d, XY = %d x %d\n",
	 bih[0], bih[1], bih[2], bih[3], xd, yd);

    printf("		L0 = %d, MX = %d, MY = %d\n",
	 l0, bih[16], bih[17]);

    printf("		Order   = %d %s%s%s%s%s\n", bih[18],
	bih[18] & JBG_HITOLO ? " HITOLO" : "",
	bih[18] & JBG_SEQ ? " SEQ" : "",
	bih[18] & JBG_ILEAVE ? " ILEAVE" : "",
	bih[18] & JBG_SMID ? " SMID" : "",
	bih[18] & 0xf0 ? " other" : "");

    printf("		Options = %d %s%s%s%s%s%s%s%s\n", bih[19],
	bih[19] & JBG_LRLTWO ? " LRLTWO" : "",
	bih[19] & JBG_VLENGTH ? " VLENGTH" : "",
	bih[19] & JBG_TPDON ? " TPDON" : "",
	bih[19] & JBG_TPBON ? " TPBON" : "",
	bih[19] & JBG_DPON ? " DPON" : "",
	bih[19] & JBG_DPPRIV ? " DPPRIV" : "",
	bih[19] & JBG_DPLAST ? " DPLAST" : "",
	bih[19] & 0x80 ? " other" : "");
    printf("		%u stripes, %d layers, %d planes\n",
	((yd >> bih[1]) +  ((((1UL << bih[1]) - 1) & xd) != 0) + l0 - 1) / l0,
	bih[1] - bih[0], bih[2]);
}

void
proff(int curOff)
{
    if (PrintOffset)
	printf("%d:	", curOff);
    else if (PrintHexOffset)
	printf("%6x:	", curOff);
}

int		curOff = 0;
int		pageNum = 0;

void
decode(FILE *fp)
{
    DWORD	magic;
    XQX_HEADER	hdr;
    int		c;
    int		rc;
    int		i;
    char	*codestr;
    FILE	*dfp = NULL;
    int		planeNum = 4;
    int		len;
    struct jbg_dec_state	s[5];
    unsigned char	bih[20];
    int			bihlen = 0;
    int			imageCnt[5] = {0,0,0,0,0};
    int         	pn = 0;
    int         	incrY = 0;
    int			totSize = 0;
    int			imagelen = 0;
    int			startPagestatus = 0;

    /*
     * <unknown> XQX format
     */
    c = getc(fp);
    if (c == EOF)
    {
	printf("EOF on file reading header.\n");
	return;
    }
    ungetc(c, fp);
    if (c == '\033' || c == '@')
    {
	char	buf[1024];

	while (fgets(buf, sizeof(buf), fp))
	{
	    proff(curOff);
	    if (buf[0] == '\033')
	    {
		printf("\\033");
		fputs(buf+1, stdout);
	    }
	    else
		fputs(buf, stdout);
	    curOff += strlen(buf);
	    if (0) {}
	    else if (strncmp(buf, "@PJL SET PAGESTATUS=START", 23) == 0)
	    {
		startPagestatus = 2;
		++pageNum;
	    }
	    else if (strncmp(buf, "@PJL SET PAGESTATUS=END", 21) == 0)
	    {
		startPagestatus = 0;
	    }
	    else if (strncmp(buf, "@PJL SET IMAGELEN=", 18) == 0)
	    {
		imagelen = atoi(&buf[18]);
		debug(2, "imagelen=%d\n", imagelen);
		break;
		proff(curOff);
		buf[51] = 0;
		printf("%s\n", buf);
		curOff += 43;
		proff(curOff);
		printf("\\033%s\n", buf+44);
		curOff += 9;
		break;
	    }
	}
	if (feof(fp))
	    return;
    }

    proff(curOff);
    printf("DDST_JBIG_DATA_BEGIN %d bytes\n", imagelen);
    for (i = 0; i < imagelen; )
    {
	unsigned char buf[4024*1024];
//printf("fread\n");
	rc = fread(buf, 1, imagelen, fp);
	if (i == 0 && startPagestatus == 2)
	{
	    memcpy(bih, buf, bihlen = 20);
//printf("data: %x\n", buf[3]);
	    if (DecFile && startPagestatus == 2)
	    {
		size_t	cnt;

		startPagestatus = 1;
		jbg_dec_init(&s[pn]);
		rc = jbg_dec_in(&s[pn], bih, bihlen, &cnt);
		if (rc == JBG_EIMPL)
		    error(1, "JBIG uses unimplemented feature\n");
	    }
	    print_bih(bih);
	}
	i += rc;
	curOff += rc;
rc = 0;
	if (rc == 0)
	{
	    debug(1, "imagelen = %d startPagestatus = %d\n",
		imagelen, startPagestatus);
	    for (i = 0; i < imagelen; ++i)
	    {
//printf("data: %x\n", buf[3]);
		c = buf[i];
		if (DecFile)
		{
		    size_t		cnt;
		    unsigned char	ch = c;

		    rc = JBG_EAGAIN;
		    rc = jbg_dec_in(&s[pn], &ch, 1, &cnt);
//printf("in: rc=%d ch=%x cnt=%ld\n", rc, ch, cnt);
		    if (rc == JBG_EOK)
		    {
			int	h, w, len;
			unsigned char *image;

//printf("JBG_OK!\n");
			// debug(0, "JBG_EOK: %d\n", pn);
			h = jbg_dec_getheight(&s[pn]);
			w = jbg_dec_getwidth(&s[pn]);
			image = jbg_dec_getimage(&s[pn], 0);
			len = jbg_dec_getsize(&s[pn]);
			//debug(0, "OK image len = %d\n", len);
			if (image)
			{
			    char	buf[512];
			    sprintf(buf, "%s-%02d-%d.pbm",
				    DecFile, pageNum, planeNum);
			    dfp = fopen(buf,
					imageCnt[planeNum] ? "a" : "w");
			    if (dfp)
			    {
				if (imageCnt[planeNum] == 0)
				    fprintf(dfp, "P4\n%8d %8d\n", w, h);
				imageCnt[planeNum] += incrY;
				rc = fwrite(image, 1, len, dfp);
				fclose(dfp);
			    }
			}
			else
			    debug(0, "Missing image %dx%d!\n", h, w);
			jbg_dec_free(&s[pn]);
		    }
		}
	    }
	    break;
	}
    }
    proff(curOff);
    printf("DDST_JBIG_DATA_END\n");
    return;
    
    /*
     * ??? XQX_MAGIC format
     */
    rc = fread(&magic, len = sizeof(magic), 1, fp);
    if (rc != 1)
    {
	printf("Missing XQX Magic number\n");
	return;
    }

    proff(curOff);
    printf("XQX_MAGIC, 0x%lx (%.4s)\n", (long)magic, (char *) &magic);

    if (memcmp((char *) &magic, ",XQX", 4))
    {
	printf("	Don't understand magic number 0x%lx\n", (long)magic);
	return;
    }

    curOff += len;
    for (;;)
    {
	proff(curOff);

	rc = fread(&hdr, len = sizeof(hdr), 1, fp);
	if (rc != 1) break;
	curOff += len;

	hdr.type = be32(hdr.type);
	hdr.items = be32(hdr.items);

	#define	CODESTR(X) case X: codestr = #X;
	switch (hdr.type)
	{
	    CODESTR(XQX_START_DOC)	break;
	    CODESTR(XQX_END_DOC)	break;
	    CODESTR(XQX_START_PAGE)	++pageNum; break;
	    CODESTR(XQX_END_PAGE)	break;
	    CODESTR(XQX_START_PLANE)	break;
	    CODESTR(XQX_END_PLANE)	break;
	    CODESTR(XQX_JBIG)		break;
	    default:			codestr = NULL; break;
	}

	if (codestr)
	    printf("%s(%ld), %ld items",
		    codestr, (long) hdr.type, (long) hdr.items);
	else
	    printf("XQX_0x%lx, %ld items", (long) hdr.type, (long) hdr.items);

	if (hdr.type == XQX_START_PAGE)
	    printf(" [Page %d]", pageNum);
	printf("\n");

	if (hdr.type == XQX_JBIG)
	{
	    for (i = 0; i < hdr.items; ++i)
	    {
		c = fgetc(fp);
		if (DecFile)
		{
		    size_t		cnt;
		    unsigned char	ch = c;

		    rc = JBG_EAGAIN;
		    rc = jbg_dec_in(&s[pn], &ch, 1, &cnt);
		    if (rc == JBG_EOK)
		    {
			int	h, w, len;
			unsigned char *image;

			// debug(0, "JBG_EOK: %d\n", pn);
			h = jbg_dec_getheight(&s[pn]);
			w = jbg_dec_getwidth(&s[pn]);
			image = jbg_dec_getimage(&s[pn], 0);
			len = jbg_dec_getsize(&s[pn]);
			if (image)
			{
			    char	buf[512];
			    sprintf(buf, "%s-%02d-%d.pbm",
				    DecFile, pageNum, planeNum);
			    dfp = fopen(buf,
					imageCnt[planeNum] ? "a" : "w");
			    if (dfp)
			    {
				if (imageCnt[planeNum] == 0)
				    fprintf(dfp, "P4\n%8d %8d\n", w, h);
				imageCnt[planeNum] += incrY;
				rc = fwrite(image, 1, len, dfp);
				fclose(dfp);
			    }
			}
			else
			    debug(0, "Missing image %dx%d!\n", h, w);
			jbg_dec_free(&s[pn]);
		    }
		}
	    }
	    curOff += hdr.items;
	    totSize += hdr.items;
	}
	else if (hdr.type == 2 && hdr.items == 0)
	    break;
	else
	{
	    for (i = 0; i < hdr.items; ++i)
	    {
		XQX_ITEM_HEADER	item;
		DWORD		val;
		int		j;

		proff(curOff);
		rc = fread(&item, len = sizeof(item), 1, fp);
		if (rc != 1) break;
		curOff += len;

		item.type = be32(item.type);
		item.size = be32(item.size);
		switch (item.type)
		{
		    CODESTR(XQXI_DMDUPLEX)		break;
		    CODESTR(XQXI_DMDEFAULTSOURCE)	break;
		    CODESTR(XQXI_DMMEDIATYPE)		break;
		    CODESTR(XQXI_RESOLUTION_X)		break;
		    CODESTR(XQXI_RESOLUTION_Y)		break;
		    CODESTR(XQXI_RASTER_X)		break;
		    CODESTR(XQXI_RASTER_Y)		break;
		    CODESTR(XQXI_VIDEO_BPP)		break;
		    CODESTR(XQXI_VIDEO_X)		break;
		    CODESTR(XQXI_VIDEO_Y)		break;
		    CODESTR(XQXI_ECONOMODE)		break;
		    CODESTR(XQXI_DMPAPER)		break;
		    CODESTR(XQXI_DUPLEX_PAUSE)		break;
		    CODESTR(XQXI_BIH)			break;
		    CODESTR(XQXI_END)			break;
		    default:			codestr = NULL; break;
		}
		if (item.size == 4)
		{
		    rc = fread(&val, len = sizeof(val), 1, fp);
		    if (rc != 1) break;
		    val = be32(val);
		    if (codestr)
			printf("	%s, %ld (0x%lx)",
				    codestr, (long) val, (long) val);
		    else
			printf("	XQXI_0x%x, %ld (0x%lx)",
				    item.type, (long) val, (long) val);
		}
		else if (item.size == 20)
		{
		    rc = fread(bih, bihlen = sizeof(bih), 1, fp);
		    if (rc != 1) break;
		    printf("	%s(0x%lx)\n", codestr, (long) item.type);
		    print_bih(bih);
		    if (DecFile)
		    {
			size_t	cnt;

			jbg_dec_init(&s[pn]);
			rc = jbg_dec_in(&s[pn], bih, bihlen, &cnt);
			if (rc == JBG_EIMPL)
			    error(1, "JBIG uses unimplemented feature\n");
		    }
		}
		else
		{
		    printf("	XQXI_0x%lx, %ld size,",
			    (long) item.type, (long) item.size);

		    for (j = 0; j < item.size; ++j)
		    {
			c = fgetc(fp);
			printf(" %02x" , c);
		    }
		}
		curOff += item.size;
		printf("\n");
	    }
	}
    }
    printf("Total size: %d bytes\n", totSize);
}

int
main(int argc, char *argv[])
{
	extern int	optind;
	extern char	*optarg;
	int		c;

	while ( (c = getopt(argc, argv, "d:hoD:?h")) != EOF)
		switch (c)
		{
		case 'd': DecFile = optarg; break;
		case 'o': PrintOffset = 1; break;
		case 'h': PrintHexOffset = 1; break;
		case 'D': Debug = atoi(optarg); break;
		default: usage(); exit(1);
		}

	argc -= optind;
	argv += optind;

	if (argc > 0)
        {
            FILE        *fp;

	    fp = fopen(argv[0], "r");
	    if (!fp)
		error(1, "file '%s' doesn't exist\n", argv[0]);
	    for (;;)
	    {
		decode(fp);
		c = getc(fp); ungetc(c, fp);
                if (feof(fp))
                    break;
	    }
	    fclose(fp);
        }
        else
	{
	    for(;;)
	    {
		decode(stdin);
		c = getc(stdin); ungetc(c, stdin);
		if (feof(stdin))
		    break;
	    }
	}
	printf("\n");

	exit(0);
}
